File: Solutions/1339. Maximum Product of Splitted Binary Tree/1339. Maximum Product of Splitted Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private long totalsum=0,res=0;
    public int maxProduct(TreeNode root) {
        totalsum=dfs(root);
        dfs(root);
        return (int)(res%(1e9 + 7));
    }
    int dfs(TreeNode root){
        if(root==null)
            return 0;
        int cursum=dfs(root.left)+dfs(root.right)+root.val;
        res=Math.max(res,cursum*(totalsum-cursum));
        return cursum;
    }
}

File: Solutions/1339. Maximum Product of Splitted Binary Tree/1339. Maximum Product of Splitted Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    long res=0,totalsum=0;
    int mod=1e9+7;
    int maxProduct(TreeNode* root) {
        totalsum=dfs(root);
        dfs(root);
        return res%mod;
    }
    int dfs(TreeNode* root){
        if(!root)
            return 0;
        int cursum=dfs(root->left)+dfs(root->right)+root->val;
        res=max(res,(totalsum-cursum)*cursum);
        return cursum;
    }
};